File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {

    Map<Node, Node> visited = new HashMap<>();        

    public Node cloneGraph(Node node) {

        if(node == null){
            return null;
        }

        copyGraph(node);

        return visited.get(node);
    }

    public Node copyGraph(Node node){
        Node newNode = new Node(node.val);
        visited.put(node, newNode);
        for(Node neighbor : node.neighbors){
            if(!visited.containsKey(neighbor)){
                visited.put(neighbor, copyGraph(neighbor));
            }
            newNode.neighbors.add(visited.get(neighbor));
        }
        return newNode;
    }
}

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:

        # Time : O(V+E)
        # Space : O(V+E)        

        if node is None:
            return
        
        q = deque() # O(1)
        q.append(node) # O(1)
        nodeMap = {node.val : Node(node.val)} # O(1)

        while q: # O(V)
            currNode = q.popleft()
            newNode = nodeMap[currNode.val]
            for neighbor in currNode.neighbors: #O(E)
                if neighbor.val not in nodeMap:
                    nodeMap[neighbor.val] = Node(neighbor.val)
                    q.append(neighbor)
                newNode.neighbors.append(nodeMap[neighbor.val])

        
        return nodeMap[node.val]